### Avance 1: EDA (Análisis exploratorio de datos)

In [1]:
#primero vamos a descargar e instalar las librerias que pide usar el trabajo integrador

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Ahora vamos a leer el archivo base desde Google Drive
# El enlace es: https://drive.google.com/file/d/1VMmmmmXpOsA_nm220fVqM4AVtE-9NjuJ/view?usp=sharing

# Extraemos el ID del archivo de la URL de Google Drive
file_id = "1VMmmmmXpOsA_nm220fVqM4AVtE-9NjuJ"

# Creamos la URL directa de descarga para Google Drive
download_url = f"https://drive.google.com/uc?id={file_id}&export=download"

# Leemos el archivo CSV directamente desde la URL de descarga
df = pd.read_csv(download_url)

# Verificamos que se haya cargado correctamente
print(f"Dataset cargado correctamente. Dimensiones: {df.shape}")
print("\nPrimeras 5 filas:")
print(df.head())

Dataset cargado correctamente. Dimensiones: (30000, 17)

Primeras 5 filas:
   id_persona    nombre apellido  edad     genero ciudad_residencia  \
0  2550327378   Jackson    Gomez  31.0  Masculino             Miami   
1  9446112038  Samantha     Soto  40.0   Femenino            Denver   
2  3098363243     Terry    Adams  62.0   Femenino            Denver   
3  4013002847     James  Shannon  41.0  Masculino            Boston   
4  7372911048     Susan    Jones  49.0   Femenino         San Diego   

  estrato_socioeconomico  frecuencia_visita  promedio_gasto_comida ocio  \
0                   Alto                  6                  67.51   Sí   
1                  Medio                  2                  44.92   Sí   
2                   Bajo                  2                   9.24   Sí   
3                   Alto                  5                  30.74   Sí   
4                   Bajo                  0                   0.00   No   

  consume_licor preferencias_alimenticias membr

# Ahora comenzamos con una exploracion inicial

In [3]:
df.columns  # Quiero ver los encabezados, para eso pido los nombres de columnas

Index(['id_persona', 'nombre', 'apellido', 'edad', 'genero',
       'ciudad_residencia', 'estrato_socioeconomico', 'frecuencia_visita',
       'promedio_gasto_comida', 'ocio', 'consume_licor',
       'preferencias_alimenticias', 'membresia_premium', 'telefono_contacto',
       'correo_electronico', 'tipo_de_pago_mas_usado', 'ingresos_mensuales'],
      dtype='object')

In [4]:
# Eliminamos la Columna contador 

if 'CONTADOR' in df.columns:
    df.drop(columns=['CONTADOR'], inplace=True)
    print("Columna 'CONTADOR' eliminada correctamente")
else:
    print("No se encontró la columna 'CONTADOR'")

# Verificamos que se haya cargado correctamente
print(f"\nDataset cargado correctamente. Dimensiones: {df.shape}")
print(f"Columnas disponibles: {list(df.columns)}")
print("\nPrimeras 5 filas:")
print(df.head())
df.head() #veamos las primeras filas, para entender que tipo de data me trae


No se encontró la columna 'CONTADOR'

Dataset cargado correctamente. Dimensiones: (30000, 17)
Columnas disponibles: ['id_persona', 'nombre', 'apellido', 'edad', 'genero', 'ciudad_residencia', 'estrato_socioeconomico', 'frecuencia_visita', 'promedio_gasto_comida', 'ocio', 'consume_licor', 'preferencias_alimenticias', 'membresia_premium', 'telefono_contacto', 'correo_electronico', 'tipo_de_pago_mas_usado', 'ingresos_mensuales']

Primeras 5 filas:
   id_persona    nombre apellido  edad     genero ciudad_residencia  \
0  2550327378   Jackson    Gomez  31.0  Masculino             Miami   
1  9446112038  Samantha     Soto  40.0   Femenino            Denver   
2  3098363243     Terry    Adams  62.0   Femenino            Denver   
3  4013002847     James  Shannon  41.0  Masculino            Boston   
4  7372911048     Susan    Jones  49.0   Femenino         San Diego   

  estrato_socioeconomico  frecuencia_visita  promedio_gasto_comida ocio  \
0                   Alto                  6      

,id_persona,nombre,apellido,edad,genero,ciudad_residencia,estrato_socioeconomico,frecuencia_visita,promedio_gasto_comida,ocio,consume_licor,preferencias_alimenticias,membresia_premium,telefono_contacto,correo_electronico,tipo_de_pago_mas_usado,ingresos_mensuales
0,2550327378,Jackson,Gomez,31.0,Masculino,Miami,Alto,6,67.51,Sí,No,Vegetariano,Sí,(830)220-1926,NaN,Efectivo,6425
1,9446112038,Samantha,Soto,40.0,Femenino,Denver,Medio,2,44.92,Sí,Sí,Mariscos,No,881-476-1426,NaN,Efectivo,2374
2,3098363243,Terry,Adams,62.0,Femenino,Denver,Bajo,2,9.24,Sí,Sí,Vegetariano,No,NaN,diana74@example.net,Efectivo,1110
3,4013002847,James,Shannon,41.0,Masculino,Boston,Alto,5,30.74,Sí,Sí,Carnes,Sí,NaN,scottfrey@example.com,Tarjeta,6931
4,7372911048,Susan,Jones,49.0,Femenino,San Diego,Bajo,0,0.00,No,No,Carnes,No,243.248.8919,glassgary@example.org,Tarjeta,1350


In [5]:
df.info()  # Tipos de datos y valores nulos

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 17 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   id_persona                 30000 non-null  int64  
 1   nombre                     30000 non-null  object 
 2   apellido                   30000 non-null  object 
 3   edad                       29899 non-null  float64
 4   genero                     30000 non-null  object 
 5   ciudad_residencia          30000 non-null  object 
 6   estrato_socioeconomico     30000 non-null  object 
 7   frecuencia_visita          30000 non-null  int64  
 8   promedio_gasto_comida      29855 non-null  float64
 9   ocio                       30000 non-null  object 
 10  consume_licor              30000 non-null  object 
 11  preferencias_alimenticias  28597 non-null  object 
 12  membresia_premium          30000 non-null  object 
 13  telefono_contacto          14834 non-null  obj

### Esta info inicial, nos cuenta que el archivo CSV:
### 30 mil Filas (registros de clientes) y 17 columnas que nos cuentan las carateristicas de estos clientes.
### Tipo de datos: 3 columnas de números enteros, 2 columas de numeros decimales y 12 son object (texto a revisar) 
### Por la diferencia entre las 30mil filas y las que contienen valor segun cada columna, detectamos que hay valores nulos

In [6]:
df.describe()

,id_persona,edad,frecuencia_visita,promedio_gasto_comida,ingresos_mensuales
count,3.000000e+04,29899.000000,30000.000000,29855.000000,30000.000000
mean,5.504765e+09,49.665006,3.896133,32.603452,5389.755867
std,2.602799e+09,23.839550,2.741532,26.402601,4538.491728
min,1.000153e+09,-5.000000,-3.000000,0.000000,800.000000
25%,3.243617e+09,33.000000,2.000000,13.290000,1860.000000
50%,5.515865e+09,49.000000,4.000000,25.510000,3402.000000
75%,7.754426e+09,65.000000,5.000000,44.400000,7761.000000
max,9.999627e+09,300.000000,10.000000,149.970000,17999.000000


## Resumen estadistico Basico
### Segun este punto:
### Edad negativa y muy longeva (300), por lo que hay que revisar estos datos.
### Los gastos en comida, tiene un valor Cero como minimo, lo que tambien se debe revisar.

 -----------------------
 ## Para no trabajar con TODA la base de datos vamos a filtrar por país

In [7]:
df_chicago = df[df['ciudad_residencia'] == 'Chicago'].copy()   #filtramosy hacemos copia para no afectar el original

In [8]:
print(f"Registros en Chicago: {df_chicago.shape[0]} filas, {df_chicago.shape[1]} columnas") #para Chequear que el filtrado esta ok

Registros en Chicago: 5384 filas, 17 columnas


### Comenzamos la limpieza de datos con NAN (Valores Nulos)

In [9]:
# Contar los valores nulos por columna
print("Valores nulos:\n", df_chicago.isna().sum())

Valores nulos:
 id_persona                      0
nombre                          0
apellido                        0
edad                           23
genero                          0
ciudad_residencia               0
estrato_socioeconomico          0
frecuencia_visita               0
promedio_gasto_comida          35
ocio                            0
consume_licor                   0
preferencias_alimenticias     268
membresia_premium               0
telefono_contacto            2730
correo_electronico           2712
tipo_de_pago_mas_usado          0
ingresos_mensuales              0
dtype: int64


In [10]:
#mejoremos la visual, quedandonos con los valores a trabajar
print("\nColumnas con nulos:")
print(df_chicago.isna().sum()[df_chicago.isna().sum() > 0])


Columnas con nulos:
edad                           23
promedio_gasto_comida          35
preferencias_alimenticias     268
telefono_contacto            2730
correo_electronico           2712
dtype: int64


## Empezemos con el tratamiento de los datos faltantes.

### EDAD

In [11]:
#Veamos que valores atípicos aun tenemos en la edad 

# Primero, analicemos la distribución actual de la edad
print("Análisis inicial de la columna 'edad':")
print(f"Rango de edad: {df_chicago['edad'].min()} - {df_chicago['edad'].max()}")
print(f"Media: {df_chicago['edad'].mean():.2f}")
print(f"Mediana: {df_chicago['edad'].median()}")

Análisis inicial de la columna 'edad':
Rango de edad: -5.0 - 300.0
Media: 49.18
Mediana: 49.0


In [12]:
# Identificar valores problematicos
edades_negativas = df_chicago[df_chicago['edad'] < 0]
edades_extremas = df_chicago[df_chicago['edad'] > 100]

print(f"\nRegistros con edad negativa: {len(edades_negativas)}")
print(f"Registros con edad > 100: {len(edades_extremas)}")


Registros con edad negativa: 16
Registros con edad > 100: 13


In [13]:
# Definir limites razonables para la edad (5 a 90 años) de alguien que visitaria un restaurante
edad_min_valida = 5
edad_max_valida = 90

# Filtrar solo las edades delimitadas
df_chicago_clean = df_chicago[
    (df_chicago['edad'] >= edad_min_valida) & 
    (df_chicago['edad'] <= edad_max_valida)
].copy()

print(f"\nRegistros después de limpiar edades atípicas: {df_chicago_clean.shape[0]}")
print(f"Registros eliminados: {len(df_chicago) - len(df_chicago_clean)}")



Registros después de limpiar edades atípicas: 5332
Registros eliminados: 52


In [14]:
## Veamos como quedo la media y la mediana
print(f"Media: {df_chicago_clean['edad'].mean():.2f}")
print(f"Mediana: {df_chicago_clean['edad'].median()}")


Media: 48.73
Mediana: 49.0


In [15]:
#Revisemos correlacion con otra variante

# Calculamos estadísticas por género 
estadisticas_edad = df_chicago_clean.groupby('genero')['edad'].agg(['mean', 'median', 'count'])
estadisticas_edad = estadisticas_edad.round(1)  # Redondeamos a 1 decimal


print("Estadísticas de edad x genero:")
print(estadisticas_edad)

Estadísticas de edad x genero:
           mean  median  count
genero                        
Femenino   48.9    48.0   2649
Masculino  48.6    49.0   2683


In [16]:
# Las diferencias entre media y mediana, por genero difieren en un punto

#Rellenamos con la mediana por genero

# Calcular la mediana por género
mediana_por_genero = df_chicago_clean.groupby('genero')['edad'].transform('median')

# Rellenar nulos
df_chicago_clean['edad'] = df_chicago_clean['edad'].fillna(mediana_por_genero)

# Verificar
print("Valores nulos en 'edad' después del relleno:", df_chicago_clean['edad'].isnull().sum())

Valores nulos en 'edad' después del relleno: 0


#### Vamos con Promedio Gastos Comida, ya que se trata de un promedio, busquemos que es mejor si la media o la mediana

In [17]:
# Sigamos con la columna promedi_gasto_comida
print(df_chicago_clean['promedio_gasto_comida'].describe())


count    5297.000000
mean       25.563687
std        22.092826
min         0.000000
25%         8.730000
50%        19.090000
75%        34.940000
max       104.920000
Name: promedio_gasto_comida, dtype: float64


In [18]:
# Verifiquemos si tenemos registros de promedio_gasto_comida = 0
gasto_cero = df_chicago_clean[df_chicago_clean['promedio_gasto_comida'] == 0]

# Distribución de visitas para esos registros
print("Distribución de frecuencia_visita para registros con gasto $0:")
print(gasto_cero['frecuencia_visita'].value_counts().sort_index())

# Identificar cuáles tienen visitas 0
sin_visitas = gasto_cero[gasto_cero['frecuencia_visita'] == 0]
print(f"\nRegistros con gasto $0 y sin visitas: {len(sin_visitas)}")


Distribución de frecuencia_visita para registros con gasto $0:
frecuencia_visita
0    260
Name: count, dtype: int64

Registros con gasto $0 y sin visitas: 260


### Esto nos muestra que los 260 registros con gasto $0 también tienen 0 visitas.
#### Esto significa que los datos son coherentes, probablemente sean usuarios inactivos (nunca visitaron ni consumieron)
#### Se procede a eliminarlos para analizar solo clientes activos

In [19]:
# Eliminar registros donde gasto = 0 Y frecuencia = 0
before = len(df_chicago_clean)

df_chicago_clean = df_chicago_clean[
    (df_chicago_clean['promedio_gasto_comida'] > 0) | 
    (df_chicago_clean['frecuencia_visita'] > 0)
].copy()

# Verificación
after = len(df_chicago_clean)
print(f"Registros eliminados (gasto $0 Y frecuencia 0): {before - after}")
print(f"Nuevo total de registros: {after}")

# Comprobar que no quedaron ceros
gasto_cero = df_chicago_clean[df_chicago_clean['promedio_gasto_comida'] == 0]
frecuencia_cero = df_chicago_clean[df_chicago_clean['frecuencia_visita'] == 0]

print("\nDespués de la limpieza:")
print(f"Registros con gasto $0: {len(gasto_cero)}")
print(f"Registros con frecuencia 0: {len(frecuencia_cero)}")

Registros eliminados (gasto $0 Y frecuencia 0): 263
Nuevo total de registros: 5069

Después de la limpieza:
Registros con gasto $0: 0
Registros con frecuencia 0: 0


### Y que pasa con las inconsistencias en cantidad de visitas

In [20]:
# Revisemos que tipo de visitas quedan por corregir
# Definir condiciones a revisar
condiciones = {
    "Visitas = 0": df_chicago_clean['frecuencia_visita'] == 0,
    "Visitas negativas": df_chicago_clean['frecuencia_visita'] < 0
}

# Recorrer y mostrar resultados
for nombre, condicion in condiciones.items():
    registros = df_chicago_clean[condicion]
    print(f"{nombre}: {len(registros)} registros")
    print("Distribución de frecuencia_visita:")
    print(registros['frecuencia_visita'].value_counts().sort_index())
    print("-" * 40)

Visitas = 0: 0 registros
Distribución de frecuencia_visita:
Series([], Name: count, dtype: int64)
----------------------------------------
Visitas negativas: 281 registros
Distribución de frecuencia_visita:
frecuencia_visita
-3    281
Name: count, dtype: int64
----------------------------------------


### En el dataset se identificaron registros con valores problemáticos en la columna frecuencia_visita:
Valores negativos (-3): es probable que sea un error de registro. Si los registros tienen datos completos, como promedio_gasto_comida distinto de $0, id, nombre, apellido, género y al menos teléfono o correo, se consideran clientes válidos.

Se decidió:
No borrar registros.
Corregir los negativos tomando el valor absoluto (-3 → 3).

In [21]:
#Procedemos a corregir los valores negativos (Error de tipeo: -3 -> 3)

# Identificar registros con frecuencia negativa
negativos = df_chicago_clean[df_chicago_clean['frecuencia_visita'] < 0]
print(f"Registros con frecuencia negativa (error): {len(negativos)}")

if len(negativos) > 0:
    # Aplicar valor absoluto para corregir el error
    df_chicago_clean.loc[negativos.index, 'frecuencia_visita'] = negativos['frecuencia_visita'].abs()
    print(f"Valores negativos corregidos: {len(negativos)} registros (ej: -3 → 3)")
else:
    print("No se encontraron valores negativos.")

# Verificamos
print(f"\nDistribucion final de 'frecuencia_visita':")
distribucion_final = df_chicago_clean['frecuencia_visita'].value_counts().sort_index()
print(distribucion_final)

Registros con frecuencia negativa (error): 281
Valores negativos corregidos: 281 registros (ej: -3 → 3)

Distribucion final de 'frecuencia_visita':
frecuencia_visita
1      268
2      644
3      986
4      909
5     1048
6      640
7      163
8      123
9      140
10     148
Name: count, dtype: int64



### Revisemos si aun nos quedan valores nulos de Gasto promedio en cuyo caso los rellenamos con la mediana

In [22]:
# Ver valores nulos restantes
nulos_gasto = df_chicago_clean['promedio_gasto_comida'].isnull().sum()
print(f"Valores nulos en promedio_gasto_comida: {nulos_gasto}")

# Calcular y usar la mediana directamente
mediana_gasto = df_chicago_clean['promedio_gasto_comida'].median()
print(f"Mediana del gasto: ${mediana_gasto:.2f}")

Valores nulos en promedio_gasto_comida: 32
Mediana del gasto: $20.32


In [23]:
# Revisemos que mas nos queda

valores_nulos_final = df_chicago_clean.isnull().sum()
valores_nulos_final = valores_nulos_final[valores_nulos_final > 0]

if len(valores_nulos_final) == 0:
    print("No hay valores nulos en el dataset")
else:
    print("Valores nulos restantes:")
    print(valores_nulos_final)

Valores nulos restantes:
promedio_gasto_comida          32
preferencias_alimenticias     255
telefono_contacto            2584
correo_electronico           2547
dtype: int64


In [24]:
# Vamos a ver que esta pasando con las Preferencias alimenticias

# Ver valores nulos
nulos_preferencias = df_chicago_clean['preferencias_alimenticias'].isnull().sum()
print(f"Valores nulos en preferencias_alimenticias: {nulos_preferencias}")

# Ver la distribucion actual de preferencias
print("\nDistribucion actual de preferencias alimenticias:")
print(df_chicago_clean['preferencias_alimenticias'].value_counts())


Valores nulos en preferencias_alimenticias: 255

Distribucion actual de preferencias alimenticias:
preferencias_alimenticias
Carnes         1499
Vegetariano    1057
Mariscos        664
Otro            549
Vegano          531
Pescado         514
Name: count, dtype: int64


In [25]:
# Para entender mejor que imputar, vamos con un analisis cruzado de datos
#Revisemos por genero
print("Distribucion de preferencias por genero:")
preferencias_por_genero = pd.crosstab(df_chicago_clean['genero'], 
                                     df_chicago_clean['preferencias_alimenticias'])
print(preferencias_por_genero)


Distribucion de preferencias por genero:
preferencias_alimenticias  Carnes  Mariscos  Otro  Pescado  Vegano  \
genero                                                               
Femenino                      760       319   257      243     265   
Masculino                     739       345   292      271     266   

preferencias_alimenticias  Vegetariano  
genero                                  
Femenino                           534  
Masculino                          523  


In [26]:
# Ahora analicemos por grupos de edad
# Crear grupos de edad
df_chicago_clean['grupo_edad'] = pd.cut(df_chicago_clean['edad'], 
                                         bins=[0, 25, 35, 45, 55, 100], 
                                         labels=['18-25', '26-35', '36-45', '46-55', '56+'])

print("\nDistribucion de preferencias por grupo de edad:")
preferencias_por_edad = pd.crosstab(df_chicago_clean['grupo_edad'], 
                                   df_chicago_clean['preferencias_alimenticias'])
print(preferencias_por_edad)


Distribucion de preferencias por grupo de edad:
preferencias_alimenticias  Carnes  Mariscos  Otro  Pescado  Vegano  \
grupo_edad                                                           
18-25                         190        77    61       62      69   
26-35                         271       121   100       88     100   
36-45                         229       105   101       72      80   
46-55                         231       100    78       84      90   
56+                           578       261   209      208     192   

preferencias_alimenticias  Vegetariano  
grupo_edad                              
18-25                              149  
26-35                              152  
36-45                              179  
46-55                              156  
56+                                421  


In [27]:
# Estrategia de imputacien inteligente basada en genero y grupo de edad

# Primero creemos grupos de edad mes detallados
df_chicago_clean['grupo_edad'] = pd.cut(df_chicago_clean['edad'], 
                                         bins=[0, 25, 35, 45, 55, 65, 100], 
                                         labels=['18-25', '26-35', '36-45', '46-55', '56-65', '66+'])

# Funcien para imputar basado en genero y grupo de edad
def imputar_preferencia_inteligente(fila):
    if pd.isna(fila['preferencias_alimenticias']):
        try:
            # Buscar personas del mismo genero y grupo de edad
            mismo_grupo = df_chicago_clean[
                (df_chicago_clean['genero'] == fila['genero']) & 
                (df_chicago_clean['grupo_edad'] == fila['grupo_edad'])
            ]
            
            if len(mismo_grupo) > 0:
                # Calcular la moda del subgrupo
                moda_grupo = mismo_grupo['preferencias_alimenticias'].mode()
                if len(moda_grupo) > 0:
                    return moda_grupo[0]
            
            # Si no hay suficientes datos en el subgrupo, usar solo genero
            mismo_genero = df_chicago_clean[df_chicago_clean['genero'] == fila['genero']]
            moda_genero = mismo_genero['preferencias_alimenticias'].mode()
            if len(moda_genero) > 0:
                return moda_genero[0]
                
        except:
            # En caso de error, usar la moda general
            pass
            
    return fila['preferencias_alimenticias']

# Aplicar la imputacion
print("Aplicando imputacion inteligente...")
nulos_antes = df_chicago_clean['preferencias_alimenticias'].isnull().sum()

df_chicago_clean['preferencias_alimenticias'] = df_chicago_clean.apply(
    imputar_preferencia_inteligente, axis=1
)

nulos_despues = df_chicago_clean['preferencias_alimenticias'].isnull().sum()
print(f"Valores nulos imputados: {nulos_antes - nulos_despues}")


# Verificacion final
nulos_final = df_chicago_clean['preferencias_alimenticias'].isnull().sum()
print(f"Valores nulos finales: {nulos_final}")

# Eliminar columna temporal
df_chicago_clean = df_chicago_clean.drop('grupo_edad', axis=1)

# Mostrar distribucion final
print("\nDistribucion final de preferencias alimenticias:")
print(df_chicago_clean['preferencias_alimenticias'].value_counts())

Aplicando imputacion inteligente...


Valores nulos imputados: 255
Valores nulos finales: 0

Distribucion final de preferencias alimenticias:
preferencias_alimenticias
Carnes         1754
Vegetariano    1057
Mariscos        664
Otro            549
Vegano          531
Pescado         514
Name: count, dtype: int64


### Bueno ahora veamos si podemos entender mejor la ultimas dos columnas con valores nulos
#### Para eso vamos a ver si misma persona visito varias veces

In [28]:
# Primero creo la columna nombre_completo 
df_chicago_clean['nombre_completo'] = df_chicago_clean['nombre'].astype(str) + ' ' + df_chicago_clean['apellido'].astype(str)
df_chicago_clean = df_chicago_clean.drop(['nombre', 'apellido'], axis=1)

# Ver cuantos nombres unicos tenemos vs registros totales
nombres_unicos = df_chicago_clean['nombre_completo'].nunique()
print(f"Registros totales: {len(df_chicago_clean)}")
print(f"Nombres unicos: {nombres_unicos}")
print(f"Registros duplicados por nombre: {len(df_chicago_clean) - nombres_unicos}")


Registros totales: 5069
Nombres unicos: 4877
Registros duplicados por nombre: 192


In [29]:
# Revisemos si esas personas tiene datos faltantes

# Filtrar solo los nombres que aparecen mas de una vez
nombres_repetidos = df_chicago_clean['nombre_completo'].value_counts()
nombres_repetidos = nombres_repetidos[nombres_repetidos > 1]

# Obtener todos los registros de nombres repetidos
registros_repetidos = df_chicago_clean[df_chicago_clean['nombre_completo'].isin(nombres_repetidos.index)]

print(f"Total de registros con nombres repetidos: {len(registros_repetidos)}")
print(f"Esto representa {len(registros_repetidos) - len(nombres_repetidos)} registros duplicados")

# Analizar valores nulos en los registros repetidos
print("\n VALORES NULOS EN REGISTROS DUPLICADOS")

nulos_en_repetidos = registros_repetidos[['telefono_contacto', 'correo_electronico']].isnull().sum()
print(nulos_en_repetidos)

Total de registros con nombres repetidos: 360
Esto representa 192 registros duplicados

 VALORES NULOS EN REGISTROS DUPLICADOS
telefono_contacto     191
correo_electronico    181
dtype: int64


### OPORTUNIDADES DE COMPLETADO 
#### Hay 192 registros duplicados (360 registros en total)
#### De estos, 191 tienen teléfono nulo y 181 tienen correo nulo

In [30]:
# Funcion optimizada para completar datos
def completar_datos_grupo(grupo):
    for columna in ['telefono_contacto', 'correo_electronico']:
        valores_no_nulos = grupo[columna].dropna()
        if len(valores_no_nulos) > 0:
            grupo[columna] = grupo[columna].fillna(valores_no_nulos.iloc[0])
    return grupo

# Aplicar completado de datos
df_chicago_completado = df_chicago_clean.groupby('nombre_completo', group_keys=False).apply(completar_datos_grupo)

# Resultados concisos
nulos_despues = df_chicago_completado[['telefono_contacto', 'correo_electronico']].isnull().sum()
completados_telefono = 2584 - nulos_despues['telefono_contacto']
completados_correo = 2547 - nulos_despues['correo_electronico']

print(f"Datos completados: Telefonos: {completados_telefono} | Correos: {completados_correo}")


Datos completados: Telefonos: 99 | Correos: 85


C:\Users\yaela\AppData\Local\Temp\ipykernel_29044\2428626223.py:10: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  df_chicago_completado = df_chicago_clean.groupby('nombre_completo', group_keys=False).apply(completar_datos_grupo)


In [34]:
# Completar valores nulos restantes con "no completo"
df_chicago_completado['telefono_contacto'] = df_chicago_completado['telefono_contacto'].fillna('no completo')
df_chicago_completado['correo_electronico'] = df_chicago_completado['correo_electronico'].fillna('no completo')

# Guardar el dataset limpio y procesado en la carpeta Archivos CSV
# ../Archivos_CSV/ -> sube un nivel (del notebook) y entra en la carpeta Archivos CSV
ruta_guardado = "../Archivos_CSV/clientes_chicago_procesado.csv"
df_chicago_completado.to_csv(ruta_guardado, index=False)


In [ ]:
# Mostrar resumen de la Etapa inical del Avance 1
print("\n RESUMEN DEL PROCESAMIENTO:")
print(f"Registros iniciales en Chicago: 5,384")
print(f"Registros despues de limpieza: {len(df_chicago_completado)}")
print(f"Valores nulos normalizados: {(5384 - len(df_chicago_completado)) + completados_telefono + completados_correo}")
print(f"Valores nulos :", df_chicago_completado.isna().sum())


 RESUMEN DEL PROCESAMIENTO:
Registros iniciales en Chicago: 5,384
Registros despues de limpieza: 5069
Valores nulos normalizados: 499
Valores nulos : id_persona                    0
edad                          0
genero                        0
ciudad_residencia             0
estrato_socioeconomico        0
frecuencia_visita             0
promedio_gasto_comida        32
ocio                          0
consume_licor                 0
preferencias_alimenticias     0
membresia_premium             0
telefono_contacto             0
correo_electronico            0
tipo_de_pago_mas_usado        0
ingresos_mensuales            0
nombre_completo               0
dtype: int64


In [ ]:
# Verificar las primeras filas del dataset guardado
print("\nVista previa del dataset guardado:")
df_chicago_completado.head()


Vista previa del dataset guardado:


,id_persona,edad,genero,ciudad_residencia,estrato_socioeconomico,frecuencia_visita,promedio_gasto_comida,ocio,consume_licor,preferencias_alimenticias,membresia_premium,telefono_contacto,correo_electronico,tipo_de_pago_mas_usado,ingresos_mensuales,nombre_completo
15,3456521319,61.0,Masculino,Chicago,Medio,2,23.02,Sí,Sí,Vegetariano,No,(394)853-1965,no completo,Efectivo,2838,Richard Adams
21,2916502228,44.0,Femenino,Chicago,Bajo,3,5.07,No,Sí,Mariscos,No,no completo,no completo,App,1365,Leah Cooper
24,5736368501,54.0,Femenino,Chicago,Alto,5,34.35,No,Sí,Carnes,Sí,no completo,no completo,Tarjeta,7133,Jennifer Richardson
29,9718391309,70.0,Femenino,Chicago,Bajo,1,6.59,No,Sí,Vegetariano,No,5093254270,no completo,Tarjeta,1205,Alyssa Bryant
32,9148064715,79.0,Masculino,Chicago,Muy Alto,8,78.02,No,Sí,Pescado,Sí,no completo,no completo,App,17625,Michael Cochran


## Continuamos en el Avance 2...
